great_expectations/expectations/core/expect_query_results_to_match_source.py

```diff
@@ -0,0 +1,196 @@
+        "target_query.table",
+        "source_query.data_source_table",
+    )
+    success_keys: ClassVar[Tuple[str, ...]] = (
+        "target_query",
+        "source_data_source_name",
+        "source_query",
+        "mostly",
+    )
+    domain_keys: ClassVar[Tuple[str, ...]] = ("batch_id",)
+
+    class Config:
+        title = "Expect query results to match source"
+
+        @staticmethod
+        def schema_extra(
+            schema: Dict[str, Any], model: Type[ExpectQueryResultsToMatchSource]
+        ) -> None:
+            BatchExpectation.Config.schema_extra(schema, model)
+            schema["properties"]["metadata"]["properties"].update(
+                {
+                    "data_quality_issues": {
+                        "title": "Data Quality Issues",
+                        "type": "array",
+                        "const": DATA_QUALITY_ISSUES,
+                    },
+                    "short_description": {
+                        "title": "Short Description",
+                        "type": "string",
+                        "const": EXPECTATION_SHORT_DESCRIPTION,
+                    },
+                    "supported_data_sources": {
+                        "title": "Supported Data Sources",
+                        "type": "array",
+                        "const": SUPPORTED_DATA_SOURCES,
+                    },
+                }
+            )
+
+    @override
+    def _validate(
+        self,
+        metrics: dict,
+        runtime_configuration: dict | None = None,
+        execution_engine: ExecutionEngine | None = None,
+    ) -> Union[ExpectationValidationResult, dict]:
+        result_format: str | dict[str, Any] = self._get_result_format(
+            runtime_configuration=runtime_configuration
+        )
+
+        target_results = metrics["target_query.table"]
+        target_result_count = len(target_results)
+        source_results = metrics["source_query.data_source_table"]
+        source_result_count = len(source_results)
+
+        if target_result_count + source_result_count == 0:
+            unexpected_count = 0
+            unexpected_percent = 0.0
+        else:
+            # creates a hashmap with row values as key and count of duplicate rows as value
+            target_results_frequency_map = Counter(tuple(row.values()) for row in target_results)
+            source_results_frequency_map = Counter(tuple(row.values()) for row in source_results)
+            # decrements source row count values by target row count values
+            count_in_source_not_target = source_results_frequency_map.copy()
+            count_in_source_not_target.subtract(target_results_frequency_map)
+
+            match_count = 0
+
+            for row, subtracted_count in count_in_source_not_target.items():
+                # if row was seen more in source than in target
+                if subtracted_count == 0:
+                    match_count += source_results_frequency_map[row]
```

Comment:
Is this right? Like if source has 5, and target has 4, subtracted we get 1, which is not 0. But it is 4 matches, right? Probably tests later on will tell me if I'm misreading.

Reply:
Ah yeah you are right. So many edge cases here.